File: src/unix/apple/request.m
// This Source Code Form is subject to the terms of the MIT License.
// If a copy of the MIT License was not distributed with this file,
// You can obtain one at https://spdx.org/licenses/MIT.html.

#include "matoya.h"

#include "net-common.h"

bool MTY_HttpRequest(const char *url, const char *method, const char *headers,
	const void *body, size_t bodySize, const char *proxy, uint32_t timeout,
	void **response, size_t *responseSize, uint16_t *status)
{
	*responseSize = 0;
	*response = NULL;

	// Request
	NSMutableURLRequest *req = net_request(url, method, headers, body, bodySize, timeout);

	// Session configuration
	NSURLSessionConfiguration *cfg = net_configuration(proxy);

	// Send request
	NSURLSession *session = [NSURLSession sessionWithConfiguration:cfg];

	__block bool r = false;
	dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);

	NSURLSessionDataTask *task = [session dataTaskWithRequest:req
		completionHandler:^(NSData *data, NSURLResponse *_res, NSError *e)
	{
		NSHTTPURLResponse *res = (NSHTTPURLResponse *) _res;

		if (!data || !res || e != nil) {
			MTY_Log("NSURLConnection failed with error %d", (int32_t) [e code]);

		} else {
			*status = [res statusCode];
			*responseSize = [data length];

			if (*responseSize > 0) {
				*response = MTY_Alloc(*responseSize + 1, 1);
				[data getBytes:*response length:*responseSize];
			}

			r = true;
		}

		dispatch_semaphore_signal(semaphore);
	}];

	[task resume];

	dispatch_semaphore_wait(semaphore, dispatch_time(DISPATCH_TIME_NOW, timeout * NSEC_PER_MSEC));

	[session invalidateAndCancel];

	return r;
}

File: src/windows/request.c
// This Source Code Form is subject to the terms of the MIT License.
// If a copy of the MIT License was not distributed with this file,
// You can obtain one at https://spdx.org/licenses/MIT.html.

#include "matoya.h"

#include <stdlib.h>
#include <stdio.h>

#include "net-common.h"

bool MTY_HttpRequest(const char *url, const char *method, const char *headers,
	const void *body, size_t bodySize, const char *proxy, uint32_t timeout,
	void **response, size_t *responseSize, uint16_t *status)
{
	*responseSize = 0;
	*response = NULL;

	HINTERNET session = NULL;
	HINTERNET connect = NULL;
	HINTERNET request = NULL;

	// Parse URL
	bool r = net_connect(url, method, headers, body, bodySize, NULL, proxy, timeout, NULL, false,
		&session, &connect, &request);
	if (!r) {
		r = false;
		goto except;
	}

	// Read response headers
	r = WinHttpReceiveResponse(request, NULL);
	if (!r)
		goto except;

	// Status code query
	r = net_get_status_code(request, status);
	if (!r)
		goto except;

	// Receive response body
	while (true) {
		DWORD available = 0;
		r = WinHttpQueryDataAvailable(request, &available);
		if (!r)
			goto except;

		if (available == 0)
			break;

		// Overflow protection
		if (available > MTY_RES_MAX || *responseSize + available > MTY_RES_MAX) {
			r = false;
			goto except;
		}

		*response = MTY_Realloc(*response, *responseSize + available + 1, 1);

		DWORD read = 0;
		r = WinHttpReadData(request, (uint8_t *) *response + *responseSize, available, &read);
		if (!r)
			goto except;

		*responseSize += read;

		// Keep null character at the end of the buffer for protection
		memset((uint8_t *) *response + *responseSize, 0, 1);
	}

	except:

	if (request)
		WinHttpCloseHandle(request);

	if (connect)
		WinHttpCloseHandle(connect);

	if (session)
		WinHttpCloseHandle(session);

	if (!r) {
		MTY_Free(*response);
		*responseSize = 0;
		*response = NULL;
	}

	return r;
}
